cpp/src/text/generate_ngrams.cu

```diff
@@ -164,4 +167,122 @@ std::unique_ptr<cudf::column> generate_ngrams(cudf::strings_column_view const& s
   return detail::generate_ngrams(strings, ngrams, separator, mr);
 }
 
+namespace detail {
+namespace {
+
+struct character_ngram_generator_fn {
+  cudf::column_device_view const d_strings;
+  cudf::size_type ngrams;
+  int32_t const* d_ngram_offsets{};
+  int32_t* d_offsets{};
+  char* d_chars{};
+
+  __device__ void operator()(cudf::size_type idx)
+  {
+    if (d_strings.is_null(idx)) return;
+    auto const d_str = d_strings.element<cudf::string_view>(idx);
+    if (d_str.empty()) return;
+    auto itr          = d_str.begin();
+    auto ngram_offset = d_ngram_offsets[idx];
+    auto ngram_count  = d_ngram_offsets[idx + 1] - ngram_offset;
+    auto d_sizes      = d_offsets + ngram_offset;
+    auto out_ptr      = d_chars ? d_chars + *d_sizes : nullptr;
+    for (cudf::size_type n = 0; n < ngram_count; ++n, ++itr) {
+      auto begin = itr.byte_offset();
+      auto end   = (itr + ngrams).byte_offset();
+      if (out_ptr)
+        out_ptr =
+          cudf::strings::detail::copy_and_increment(out_ptr, d_str.data() + begin, (end - begin));
+      else
+        *d_sizes++ = end - begin;
+    }
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<cudf::column> generate_character_ngrams(cudf::strings_column_view const& strings,
+                                                        cudf::size_type ngrams,
+                                                        cudaStream_t stream,
+                                                        rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(ngrams > 1, "Parameter ngrams should be an integer value of 2 or greater");
+
+  auto strings_count = strings.size();
+  if (strings_count == 0)  // if no strings, return an empty column
+    return cudf::make_empty_column(cudf::data_type{cudf::STRING});
+
+  auto execpol        = rmm::exec_policy(stream);
+  auto strings_column = cudf::column_device_view::create(strings.parent(), stream);
+  auto d_strings      = *strings_column;
+
+  // create a vector of ngram offsets for each string
+  rmm::device_vector<cudf::size_type> ngram_offsets(strings_count + 1);
+  thrust::transform_exclusive_scan(
+    execpol->on(stream),
+    thrust::make_counting_iterator<cudf::size_type>(0),
+    thrust::make_counting_iterator<cudf::size_type>(strings_count + 1),
+    ngram_offsets.begin(),
+    [d_strings, strings_count, ngrams] __device__(auto idx) {
+      if (d_strings.is_null(idx) || (idx == strings_count)) return 0;
+      auto length = d_strings.element<cudf::string_view>(idx).length();
+      return (length < ngrams) ? 0 : (length + 1 - ngrams);
```

Comment:
Any function declared as `constexpr` can be used in device code (as per our compile options for libcudf). And `std::max()` is one of those so I will use that here instead of the `thrust::max()`.